### DISCUSSION TITLE
Controller - Service, 계층간 데이터 전달 문제

### DISCUSSION THREAD

**Discussion by yuseung0429** [BackendTheBone/MiraeCommerce#13] [General]
# Controller와 Service의 강한 결합
## Trouble
Service의 메서드 작성 중 Service의 메서드가 Return하는 값이 Controller가 클라이언트에게 전달하는 값인, Response인 것이 올바른 설계인가?

### Thinking
- Service는 Business Logic을 담당해야 한다고 생각하는데, Controller와의 강한 결합이 있으면 안된다.
- 강한 결합이 있다면, Service의 특정 메서드는 Controller에서 받는 요청만을 처리할 수 있게 되어버린다.
- 현재 소스코드에서 두 가지 형태의 메서드가 있는데, 해당 메서드들은 동일한 동작을 한다.
    1. Controller에서 받은 요청을 처리하는 메서드
    2. 다른 Service에서 받은 요청을 처리하는 메서드
- 이는 소스코드의 중복으로, 유지 보수 관점에서 좋지 않은 소스코드라고 생각한다.

### Solving
- Service는 해당 Business Logic을 수행하는데 필요한, 정형화된 포맷을 매개변수로 받고, Controller에 의존적이지 않은, 결과값을 Return한다.
- Controller는 Service에서 Return된 결과값을 통해서 Response를 만들어 사용자에게 전달한다.

### Consider
- Service의 메서드의 매개변수
    - 일단, 위 생각이 올바르다면, Controller에서 받은 Request 객체를 Service로 넘겨주는 것은 올바르지 않는 방법이라고 생각할 것이다. 
    - 두가지 방법이 있다고 생각한다.
        1. Business Logic에 필요한 매개변수를 직접 매개변수로 받는 방법
            - public ReturnObject login(String username, String password){ … }
            - 메서드에서 사용되는 필드가 늘어날 경우, 매개변수를 직접 수정해야하고, 확장에 닫혀있다.
        2. DTO 객체를 이용해서 매개변수로 받는 방법
            - public ReturnObject login(UserInfoDto userInfoDto){ … }
            - 관리해야 하는 DTO의 수가 늘어나면서 코드의 복잡성이 늘어난다.
- Service 메서드의 Return 값
    - Domain 객체가 Service 밖으로 새어나가면 어떤 Side Effect가 발생할지 모르므로, Service Layer에서 Return 하는 값은 Domain 객체가 아니라 DTO 객체여야 한다.
    - 마찬가지로 이런 방법을 사용한다면 관리해야 하는 DTO의 수가 늘어나게 된다.

## Opinion
- Service에서 받는 매개변수와 Return하는 값은 Controller와 결합을 가지면 안된다.
- Controller에서 받는 요청의 대한 객체 Nameing은 Request
- Service에서 주고 받는 매개변수와 Return하는 객체 Naming은 DTO

**Comment by hamsteak1488** [comment]
여러 방법에 대해 조사해 보았는데 저도 각 계층별 요청 및 응답 DTO를 분리하는게 더 좋아보입니다. 다만 해당 방법은 계층 간 DTO 변환 코드가 매우 많이 생기는 불편함이 있을 것 같아보입니다.

그래서 개선할 방법이 있나 찾아보았는데 ModelMapper와 MapStruct를 이용한다면 이를 해결할 수 있을 듯 합니다. 두 라이브러리 모두 필드가 같은 객체가 서로 값을 바인딩하는 작업을 간단하게 할 수 있도록 도와주는 역할을 합니다. 저는 개인적으로 MapStruct를 채택하는게 더 좋아보이는데 MapStruct는 컴파일 시점에 동작하므로 ModelMapper보다 성능을 비롯해 많은 면에서 이점을 가져오기 때문입니다.

두 라이브러리 중 하나를 적용해서 우리가 해야 하는 바인딩 작업을 간소화한다면 요청 및 응답 DTO를 추가하는데에 드는 부담을 훨씬 줄일 수 있을 것 같습니다.
